Implement compare characters length method

akaryu0206 opened the pull request:
# New Features
- Implement `length?` method (#22)
- Add tests for `length?` method

akaryu0206 committed:
[Add] Implement compare characters length method

 
- Implement `length?` method
- Add test for `length?` method

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,36 @@
+describe "[ Length Methods ]" do
+
+  describe "COMPARE CHARACTERS LENGTH" do
+    context "when the argument is String class," do
```
`the` -> `a`

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,36 @@
+
+  describe "COMPARE CHARACTERS LENGTH" do
+    context "when the argument is String class," do
+      it { expect{ "string".length?("6") }.to raise_error ArgumentError }
```
When you use `subject` and `is_expected` methods, you can write with oneline. Because it will be wrong English.

```rb
# NG
# (Wrong English: It expect XXX to equal YYY.)
it { expect(xxx).to eq(yyy) }

# OK
# (Right English: It is expected to equal YYY.)
subject { xxx }
it { is_expected.to eq yyy }
```

jagaapple replied:
In this method, it is possible to write simply using `subject` and `let` method.

```rb
let(:string) { "" }
let(:expected_length) { 0 }
subject { string.legth?(expected_length) }

context "..." do
  let(:string) { "abc" }
  let(:expected_length) { 3 }

  it { is_expected.to be true }
end

context "..." do
  let(:string) { "abcd" }
  let(:expected_length) { 3 }

  it { is_expected.to be false }
end
```

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,36 @@
+      it { expect("string".length?(5)).to be false }
+    end
+
+    context "when character length are within range of argument," do
```
`character` -> `characters`

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    context "when character length are within range of argument," do
+      it {
```
Use `do ... end` for multiline block.

jagaapple replied:
Also, specify description string for `it` method when using multiline block. In general, starting with a word `"should"` .

```rb
# OK
# (Right English: It should be true.)
it "should be true" do
  expect(xxx).to be true
end
```

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,36 @@
+      }
+    end
+
+    context "when character length are out of range of argument," do
```
`character` -> `characters`

akaryu0206 committed:
[Fix] Fix reviewed code, and improve

 
- Use `let` and `subject` in test code
- Corresponding to ruby 2.1.x to 2.3.x
- Clean code

jagaapple commented on lib/string_foundation/length.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+class String
+
+  # Compare characters length
```
Append `.` in the end of string because this is one sentence 🙄

jagaapple commented on lib/string_foundation/length.rb:
```diff
@@ -0,0 +1,17 @@
+  # Compare characters length
+  def length?(length)
+    unless [Integer, Range, Fixnum, Bignum].include?(length.class)
+      raise ArgumentError.new("argument must be Integer (it called Fixnum or Bignum in Ruby 2.3 or lower) or Range")
```
How do you feel to use `argument must be Integer (including Fixnum or Bignum) or Range` in this error message?

akaryu0206 replied:
I think it's good.
It's simple and easy to understand.

jagaapple commented on lib/string_foundation/length.rb:
```diff
@@ -0,0 +1,17 @@
+      raise ArgumentError.new("argument must be Integer (it called Fixnum or Bignum in Ruby 2.3 or lower) or Range")
+    end
+
+    return self.length == length unless length.class == Range
```
Prefer `.instance_of?()` to `.class ==` .

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    context "when a argument is String class," do
+      let(:expected_length) { "6" }
+
+      it { is_expected_as_block.to raise_error ArgumentError }
```
Wrap put parentheses `()` around for arguments like `raise_error(ArgumentError)` .

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    subject { string.length?(expected_length) }
+
+    context "when a argument is String class," do
+      let(:expected_length) { "6" }
```
Use `string.length.to_s` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length and argument are the same," do
+      let(:expected_length) { 6 }
```
Use `string.length` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length and argument are not the same," do
+      let(:expected_length) { 5 }
```
Use `string.length - 1` instead.

jagaapple replied:
Also need a test when an expected length is `string.length + 1` .

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length are within range of argument `1..10`," do
+      let(:expected_length) { 1..10 }
```
Use `(string.length - 1)..(string.length + 1)` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length are within range of argument `6..10`," do
+      let(:expected_length) { 6..10 }
```
Use `string.length..(string.length + 1)` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length are within range of argument `1..6`," do
+      let(:expected_length) { 1..6 }
```
Use `(string.length - 1)..string.length` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length are out of range of argument `1..5`," do
+      let(:expected_length) { 1..5 }
```
Use `(string.length - 2)..(string.length - 1)` instead.

jagaapple commented on spec/string_foundation/length_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    context "when characters length are out of range of argument `7..10`," do
+      let(:expected_length) { 7..10 }
```
Use `(string.length + 1)..(string.length + 2)` instead.

akaryu0206 committed:
[Clean] Refactor reviewed code

- Improve sentences
- Improve conditions
- Improve test code to abstract

Commit:
[Clean] Modify context descriptions in tests

- Modify context descriptions in tests

jagaapple reviewed (dismissed):
Good!

Commit:
[Add] Add documents for length methods

- Add documents for length methods

jagaapple reviewed (approved):
Added usage about this to a readme file.
